### ISSUE TITLE
`iox::rp::RelativePointer` violates type safety

### ISSUE THREAD

**Issue by elfenpiff** [eclipse-iceoryx/iceoryx#1121]
## Required information

It is possible to assign any arbitrary pointer to the `RelativePointer`, see the following code:
```cpp
float f;
iox::rp::RelativePointer<int> ptr{&f}; // is accepted but should result in a compile time error
int* fuu = ptr; // this causes now problems since we have a pointer to the wrong type

// as comparision
int * fuu = &f; // this leads to a compile time error.
```

The `AtomicRelocatablePointer` is not affected and works as expected.

**Comment by elfenpiff** [comment]
FYI @MatthiasKillat . I think this may be fixed quickly  - do you have time to fix this before 2.0 ?

**Comment by MatthiasKillat** [comment]
@elfenpiff I think I have the time. But at some point I wanted to also refactor the whole design (without inheritance which was a questionable decision). It would supposed to clean up some code, of course, similar to the refactoring of `relocatable_ptr` (which already happened).

Before 2.0: bugfix yes, refactoring no.
